src/wp-includes/shortcodes.php

```diff
@@ -221,6 +221,12 @@ function do_shortcode( $content, $ignore_html = false ) {
 		return $content;
 	}
 
+	$callback = static function() {
+		return 'do_shortcode';
+	};
```

Comment:
I disagree here. Not everything that is in a hook is intended to be unhooked. While hooks are great, there are things for which unhooking will simply break expected behavior, so I find the idea of requiring _every_ hook to be removable questionable.

IMO a private function is a less clean approach here than this closure.

Reply:
I usually would agree, and was a bit on the fence about this one so I'm happy to change it if needed. For now, I've updated an inline comment ([ee97a6c](https://github.com/WordPress/wordpress-develop/pull/4799/commits/ee97a6cc368f5509cd63ac840057428dee6b2892)) where that callback is added to give more context for why it's there. I agree that the intent is _not_ to have folks unhook it, but we could consider adding it to an earlier priority to ensure any other filters added overtake this one if you think it would help.

Reply:
Closure are horrible for unit test coverage reports and for profiling tools. See this example screenshot. I think closures make sense in plugins, but as a framework, WordPress does and should always, allow you to remove filters. 

Look at this profile, how can find these closures? Are they even the same closure?
<img width="1784" alt="Screenshot 2023-07-10 at 11 37 08" src="https://github.com/WordPress/wordpress-develop/assets/237508/800a6abb-33af-481e-b269-2cd19aac3a46">

Reply:
I'm wary of this turning into a broader conversation about closure usage. The closure here is tiny and introducing a private function for it instead would add things to the global scope, which is usually not a good idea for things that are private.

For profiling, I'd argue seeing the parent functions below should clarify where the closure is called.

Reply:
I agree, let's not get on the conversion about closures here. Using a closure is not appropriate here. It is not a pattern used in core and I don't want to start using it here. Let's move to a private function and have the discussion of closure elsewhere.

Reply:
Using a closure like this may also result in the same filter being added multiple files, as `_wp_filter_build_unique_id` generates a new value for everytime the closure is used. This may end up adding many many keys to the `callbacks` property of WP_Hook.

Reply:
> Using a closure like this may also result in the same filter being added multiple files, as _wp_filter_build_unique_id generates a new value for everytime the closure is used. This may end up adding many many keys to the callbacks property of WP_Hook.

But if you're passing the same variable referencing the same closure then `spl_object_hash()` returns the same hash, no? Don't wanna argue about the usefulness of closure, but pretty sure it would work.

Reply:
☝️ That's what I was thinking as well.

I'm a bit confused by the results in the screenshot in the other thread https://github.com/WordPress/wordpress-develop/pull/4799#discussion_r1258748056 🤔 Maybe this is not working like we were thinking?

Reply:
@felixarntz This screen was from a static variable like this 

```
	$callback = static function() {
		return 'do_shortcode';
	};
```

I replicate it again with commit c3776abbe4c241803332f8fad50cfc76bc387d92

<img width="701" alt="Screenshot 2023-07-10 at 22 36 10" src="https://github.com/WordPress/wordpress-develop/assets/237508/8770ed98-7868-43e0-9a57-653fe624c3cd">

Reply:
That‘s just a static closure but not a static variable

Reply:
@spacedmonkey That's not a static variable, only the closure is a static function. For it to be a static variable, it would need to be
```php
	static $callback = static function() {
		return 'do_shortcode';
	};
```